### PR TITLE
SCREAM: make test-all-scream side-effect-free

### PR DESCRIPTION
**Pull request by bartgol**
This PR makes sure the repo is returned to its original state at the end of test-all-scream execution.

This is only useful if any of the following apply:

- user tests against a baseline ref other than HEAD, and something goes wrong while switching branches, or during the baselines generation
- user does an integration test. If successful, and the branch is not ahead of master, the repo will contain a merge commit. If something goes wrong during the merge, the repo may not be clean.

This mod should not affect a normal development cycle (in particular, it has no effect when running with `--keep-tree` option).

### CODE REVIEW

File: components/scream/scripts/test_all_scream.py [hunk 1]
```diff
@@ -298,31 +300,41 @@ def run_all_tests(self):
...
+                if not success:
+                    print ("Error(s) occurred during baselines generation phase")
+
+            if self._perform_tests:
+                # Finally, run the tests
+                success &= self.run_all_tests()
+                if not success:
+                    print ("Error(s) occurred during test phase")
+
+            if not self._keep_tree:
```
Comment (jgfouca) [inline components/scream/scripts/test_all_scream.py]: This is a perfect usecase for python's `finally` semantics.

```
finally:
    if not self._keep_tree:
        # Cleanup the repo if needed, then re-throw whatever the exception was
        cleanup_repo(self._original_head, self._original_commit)
```

File: components/scream/scripts/test_all_scream.py [hunk 2]
```diff
@@ -298,31 +300,41 @@ def run_all_tests(self):
...
+                os.makedirs(test_dir)
 
-        if self._baseline_dir is None:
-            # Second, generate baselines
-            expect(self._baseline_ref is not None, "Missing baseline ref")
-
-            success = self.generate_all_baselines()
-            if not success:
-                print ("Error(s) occurred during baselines generation phase")
-                return success
```
Comment (jgfouca) [inline components/scream/scripts/test_all_scream.py]: Restore this return, it works with `finally`

File: components/scream/scripts/utils.py [hunk 3]
```diff
@@ -366,7 +366,8 @@ def merge_git_ref(git_ref, repo=None):
     """
     Merge given git ref into the current branch, and updates submodules
     """
-    run_cmd_no_fail("git merge {} -m 'Automatic merge of {}'".format(git_ref,git_ref),from_dir=repo)
+    expect(is_repo_clean(), "Cannot merge ref '{}'. The repo is not clean.".format(git_ref))
+    stat = run_cmd("git merge {} -m 'Automatic merge of {}'".format(git_ref,git_ref),from_dir=repo)
```
Comment (jgfouca) [inline components/scream/scripts/utils.py]: This should be run_cmd_no_fail I think. That, or the contents of `stat` should be checked.

File: components/scream/scripts/utils.py [hunk 4]
```diff
@@ -398,3 +399,33 @@ def checkout_git_ref(git_ref, verbose=False, repo=None):
             print("  Switched to '{}' ({})".format(git_ref,git_commit))
             print_last_commit(git_ref=git_ref)
 
+###############################################################################
+def get_git_toplevel_dir(repo=None):
+###############################################################################
+    """
+    Get repo toplevel directory
+    """
+    stat, dirname = run_cmd_no_fail("git rev-parse --show-toplevel")
```
Comment (jgfouca) [inline components/scream/scripts/utils.py]: run_cmd_no_fail does not return a stat, it only return the output and raises an expect if the command failed.

```
    return run_cmd_no_fail("git rev-parse --show-toplevel"
```

Comment (bartgol) [reply components/scream/scripts/utils.py]: True. Thanks.

File: components/scream/scripts/utils.py (hunk 3 above)
Comment (bartgol) [reply components/scream/scripts/utils.py]: True. Thanks.

File: components/scream/scripts/test_all_scream.py (hunk 1 above)
Comment (bartgol) [reply components/scream/scripts/test_all_scream.py]: Nice. Thanks!

File: components/scream/scripts/test_all_scream.py (hunk 2 above)
Comment (bartgol) [reply components/scream/scripts/test_all_scream.py]: There's a `return success` at the bottom of the function though, that should be enough, no?

**Comment by bartgol** [comment]
@jgfouca I addressed all your comments except for the `return success` one (see inline comment). I also added an extra commit to improve the processors usage in case of parallel testing.

File: components/scream/scripts/test_all_scream.py [hunk 5]
```diff
@@ -298,31 +313,43 @@ def run_all_tests(self):
...
+                if not success:
+                    print ("Error(s) occurred during baselines generation phase")
+
+            if self._perform_tests:
+                # Finally, run the tests
+                success &= self.run_all_tests()
+                if not success:
+                    print ("Error(s) occurred during test phase")
+
+        except:
```
Comment (jgfouca) [inline components/scream/scripts/test_all_scream.py]: This is not necessary I don't think. The finally alone should work.

File: components/scream/scripts/test_all_scream.py (hunk 2 above)
Comment (jgfouca) [reply components/scream/scripts/test_all_scream.py]: Yes, but the control-flow of this loop is different without the return being here.

File: components/scream/scripts/test_all_scream.py (hunk 5 above)
Comment (bartgol) [reply components/scream/scripts/test_all_scream.py]: makes sense.

Comment (bartgol) [reply components/scream/scripts/test_all_scream.py]: makes sense.

File: components/scream/scripts/test_all_scream.py (hunk 2 above)
Comment (bartgol) [reply components/scream/scripts/test_all_scream.py]: Ah, right, we want to skip the tests.

**Pull request merged**